# Heroes of Pymoli

### Setup
* Import dependencies and load/read cvs file

In [1]:
# Import dependencies
import pandas as pd

# File to Load 
pymoli = "Resources/purchase_data.csv"

# Read file and store into Pandas data frame
pymoli_df = pd.read_csv(pymoli)
# pymoli_df.head()

### Player Count

* Display the total number of players in a Data Frame

In [2]:
# Count the total number of unique players and display the result in a Pandas data frame
total_players = len(pymoli_df["SN"].unique())
total_players_df = pd.DataFrame ({"Total Players":[total_players]})

# Display the data frame
total_players_df

Total Players
0            576

### Purchasing Analysis (Total)

* Run calculations to find number of unique items, the total number of purchases made, average price per purchase and total spent on purchases

* Create a dictionary to hold the new columns and values, then create a summary Data Frame to hold the results

* Display and format the Data Frame

In [3]:
# Run calculations to find total number of unique items, the total number of purchases made, average price per purchase and total spent on purchases
unique_items = len(pymoli_df["Item ID"].unique())
number_purchases = pymoli_df["Purchase ID"].count()
total = pymoli_df["Price"].sum()
average_price = pymoli_df["Price"].mean()

# Create a dictionary to hold newly created columns and values
summary_data = {"Number of Unique Items": [unique_items], "Average Price": [average_price], "Number of Purchases": [number_purchases], "Total Revenue": [total]}

# Create a data frame of newly created columns and values
summary_df = pd.DataFrame (summary_data, columns = ["Number of Unique Items", "Average Price", "Number of Purchases", "Total Revenue"])

# Convert Average Price and Total Revenue to float, then format
summary_df["Average Price"] = summary_df["Average Price"].astype(float).map("${:,.2f}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].astype(float).map("${:,.2f}".format)

# Display the data frame
summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

### Gender Demographics

* Caclulate the percentage and count of Male Players, Female Players and Other / Non-Disclosed players
* Create, display and format the results in a Data Frame

In [4]:
# Group data by gender
gender = pymoli_df.groupby("Gender")

# Find the total count of each gender
total_gender = gender.nunique()["SN"]
                 
# Find the percentage of each gender
gender_percent = total_gender/total_players*100

# Create a new dataframe to hold calculations for percentage and count of each gender
gender_df = pd.DataFrame({"Total Count": total_gender, "Percentage of Players": gender_percent})

# Sort values in descedning order 
gender_df.sort_values(by=["Total Count"], inplace = True, ascending = False)

# Convert Percentage of Players to float, then format
gender_df["Percentage of Players"] = gender_df["Percentage of Players"].astype(float).map("{:,.2f}%".format)

# Remove the index name
gender_df.index.name = None

# Display the data frame
gender_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

### Purchasing Analysis (Gender)

* Run basic calculations to find the purchase count, average purchase price and purchase total by gender

* Create, display and format the results in a Data Frame

In [5]:
# Run calculations to find the purchase count, average purchase price, purchase total by gender and the average total purchase per person
# Purchase count per gender
purchase_count = gender["Item ID"].count()

# Total Purchase price per gender
purchase_total = gender["Price"].sum()

# Average Purchase Price per gender
average_purchase = purchase_total/purchase_count

# Avg Total Purchase per person
average_total = purchase_total/total_gender

# Create a new data frame to hold calculations for purchase count, average purchase price, purchase total by gender and the average total purchase per person
grouped_gender_df = pd.DataFrame({"Purchase Count": purchase_count, "Average Purchase Price": average_purchase, "Total Purchase Value":purchase_total, "Avg Total Purchase per Person":average_total})                 
                           
# Convert Average Purchase Price, Total Purchase Value and Avg Total Purchase per Person to float, then format
grouped_gender_df["Average Purchase Price"] = grouped_gender_df["Average Purchase Price"].astype(float).map("${:,.2f}".format)
grouped_gender_df["Total Purchase Value"] = grouped_gender_df["Total Purchase Value"].astype(float).map("${:,.2f}".format)
grouped_gender_df["Avg Total Purchase per Person"] = grouped_gender_df["Avg Total Purchase per Person"].astype(float).map("${:,.2f}".format)

# Display the data frame
grouped_gender_df.head()

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

### Age Demographics

* Establish bins for ages and categorise players accordingly

* Calculate the total number and percentages of players by age group

* Create, display and format the results in a Data Frame

In [6]:
# Figure out the minimum and maximum age 
# print(pymoli_df["Age"].max())
# print(pymoli_df["Age"].min())

# Create bins 
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9, 34.9, 39.9, 50]

# Create names for the bins
group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Slice the data and place it into bins
pymoli_df["Age Ranges"] = pd.cut(pymoli_df["Age"], bins, labels=group_labels)
pymoli_df.head()

# Create a GroupBy object based upon "Age Ranges"
age_ranges = pymoli_df.groupby("Age Ranges")

# Find how many rows fall into each bin
# print(age_df["Age"].count())

# Count total players by age range
age_count = age_ranges["SN"].nunique()
# print(total_count_age)

# Calculate percentages by age 
age_percent = (age_count/total_players)*100

# # Create a new summary data frame to hold the results 
age_df = pd.DataFrame({"Total Count": age_count,"Percentage of Players": age_percent})

# Convert Percentage of Players to float, then format
age_df["Percentage of Players"] = age_df["Percentage of Players"].astype(float).map("{:,.2f}%".format)

# Display the new summary data frame without the index name
age_df = age_df.rename_axis(None)
age_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

### Purchasing Analysis (Age)

* Run basic calculations to find purchase count, average purchase price, total purchase value and average total purcase per person, per age group

* Create, display and format the results in a Data Frame

In [7]:
# Run calculations for purchase count, average purchase price, total purchase value and average total purcase per person, per age group
# Total number of purchases per age group
age_purchase = age_ranges["Purchase ID"].count()

# Average Purchase Price per age group
age_purchase_average = age_ranges["Price"].mean()

# Total purchase value per age group
age_total_purchase = age_ranges["Price"].sum()

# Average total purcahse per person per age group
age_average_purchase = age_total_purchase/age_count

# Create a new data frame to hold calculations for purchase count, average purchase price, total purchase value and average total purcase per person, per age group
age_summary = pd.DataFrame({"Purchase Count" : age_purchase, "Average Purchase Price": age_purchase_average, "Total Purchase Value": age_total_purchase, "Avg Total Purchase per Person" : age_average_purchase})

# Convert Average Purchase Price, Total Purchase Value and Average Total Purcase per Person to float, then format
age_summary["Average Purchase Price"] = age_summary["Average Purchase Price"].astype(float).map("${:,.2f}".format)
age_summary["Total Purchase Value"] = age_summary["Total Purchase Value"].astype(float).map("${:,.2f}".format)
age_summary["Avg Total Purchase per Person"] = age_summary["Avg Total Purchase per Person"].astype(float).map("${:,.2f}".format)

# Display the data frame
age_summary

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

### Top Spenders

* Run basic calculations to identify the top 5 spenders by total purchase value; also find total purchase count and average purchase price

* Create, display and format the results in a Data Frame, ensuring to sort the data by "Total Purchase" column in descending order

In [8]:
# Group by SN
top_spenders = pymoli_df.groupby("SN")

# Run calculations to find total purchase count, average purchase price and total purchase value 
# Calculate the total purchase count
spender_purchase_count = top_spenders["Purchase ID"].count()

# Calculate average purchase price
spender_average_price = top_spenders["Price"].mean()

# Calculate total purchase value
spender_total_purchase = top_spenders["Price"].sum()

# Create a new data frame to hold calculations for total purchase count, average purchase price and total purchase value 
spender_summary_df = pd.DataFrame({"Purchase Count" : spender_purchase_count, "Average Purchase Price": spender_average_price, "Total Purchase Value": spender_total_purchase})

# Sort data frame in descending order based on purchase value 
spender_summary_df.sort_values(by=["Total Purchase Value"], inplace = True, ascending = False)

# Convert Average Purchase Price and Total Purchase Value to float, then format
spender_summary_df["Average Purchase Price"] = spender_summary_df["Average Purchase Price"].astype(float).map("${:,.2f}".format)
spender_summary_df["Total Purchase Value"] = spender_summary_df["Total Purchase Value"].astype(float).map("${:,.2f}".format)

# Display the first 5 items within the data frame
spender_summary_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

### Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns

* Group by Item ID and Item Name. Perform calculations to find top 5 most popular items and their total purchase count, item price and total purchase value 

* Create, display and format the results in a Data Frame, ensuring to sort the data by "Purchase Count" column in descending order

In [9]:
# Pull the column information from the DataFrame "pymoli_df"
items = pymoli_df[["Item ID", "Item Name", "Price"]]

# Group by item id and item name 
popular_items = items.groupby(["Item ID","Item Name"])

# Run calculations to find total purchase count, item price and total purchase value 
# Calculate the total purchase count
popular_count = popular_items["Price"].count()

# Calculate the total purchase value
popular_total = popular_items["Price"].sum()

# Calculate the price per individual item 
popular_item_price = popular_total/popular_count

# Populate data frame with obtained values
most_popular_items_df = pd.DataFrame({"Purchase Count": popular_count, 
                                   "Item Price": popular_item_price,
                                   "Total Purchase Value":popular_total})

# Sort in descending order to obtain top spender names and provide top 5 item names
popular_sorted = most_popular_items_df.sort_values("Purchase Count", ascending=False)

# Convert Item Price and Total Purchase Value to float, then format 
popular_sorted["Item Price"] = popular_sorted["Item Price"].astype(float).map("${:,.2f}".format)
popular_sorted["Total Purchase Value"] = popular_sorted["Total Purchase Value"].astype(float).map("${:,.2f}".format)

# Display the first 5 items within the data frame
popular_sorted.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

### Most Profitable Items

* Sort the Data Frame created above (Most Popular Items) total purchase value in descending order

* Display and format the results in a Data Frame

In [10]:
# Copy the Most Popular Items DataFrame and give it a new name
popular_sorted = most_popular_items_df.sort_values("Total Purchase Value", ascending=False)

# Convert Item Price and Total Purchase Value to float, then format 
popular_sorted["Item Price"] = popular_sorted["Item Price"].astype(float).map("${:,.2f}".format)
popular_sorted["Total Purchase Value"] = popular_sorted["Total Purchase Value"].astype(float).map("${:,.2f}".format)

# Display the first 5 items within the data frame
popular_sorted.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# 3 Observable Trends Based on the Data



* The majority of players in the game are males, and they are also the ones responsible for the majority of purchases within the game, out of a total 780 purchases made, 652 of those were made by a male user (83.5% of purchases), and equate to a total of $1967.64, although the average purchase price for males was not overly high which is also shown by the number of purchases made. 
 
* The majority of players are within the 20-24 age range, and account for a total of 44.79% of all players, however looking at an extended age range between 15-29, these players account for a total of 76.7% of all users meaning this game is most popular to users of the age between 15 and 29.

* The two most popular items are also the most profitable items brought within the game. Out of the most profitable items, 3 of them are also in the top 5 for most popular. These items all have a high item price, which is approximately 1.5 times (or more) higher than the average purchase price. To continue to increase profits for the game, the top 5 items need to be evaluated and marketed to the correct audience. Based on findings, this audience would be males of age range 15-29. 